samza-azure/src/main/java/org/apache/samza/NoFlushAsyncSystemProducer.java

```diff
@@ -0,0 +1,68 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import org.apache.samza.config.Config;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Helper System producer that can be used to implement system producers whose APIs don't provide flush semantics
+ * but instead provides Async callbacks e.g. Event Hubs
+ */
+public abstract class NoFlushAsyncSystemProducer extends AsyncSystemProducer {
+
+  private static final Logger LOG = LoggerFactory.getLogger(NoFlushAsyncSystemProducer.class.getName());
+  private static final long DEFAULT_FLUSH_TIMEOUT_MILLIS = Duration.ofMinutes(1L).toMillis();
+
+  public NoFlushAsyncSystemProducer(String systemName, Config config, MetricsRegistry registry) {
+    super(systemName, config, registry);
+  }
+
+  @Override
+  public synchronized void flush(String source) {
+    long incompleteSends = pendingFutures.stream().filter(x -> !x.isDone()).count();
+    LOG.info("Trying to flush pending {} sends.", incompleteSends);
+    checkCallbackThrowable("Received exception on message send.");
+    CompletableFuture<Void> future =
+        CompletableFuture.allOf(pendingFutures.toArray(new CompletableFuture[pendingFutures.size()]));
+
+    try {
+      // Block until all the pending sends are complete or timeout.
+      future.get(DEFAULT_FLUSH_TIMEOUT_MILLIS, TimeUnit.MILLISECONDS);
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      incompleteSends = pendingFutures.stream().filter(x -> !x.isDone()).count();
+      String msg = String.format("Flush failed with error. Total pending sends %d", incompleteSends);
+      LOG.error(msg, e);
+      throw new SamzaException(msg, e);
+    }
+
+    pendingFutures.clear();
```

Comment:
here you are clearing the pendingFutures, but other message could have been added to pendingFutures by the send method. This could be a race condition, where the message is removed from the pendingFutures before the callback is completed.

Suggestion, remove the message from the pendingFutures after the callback is done (in the closure in the send method), or make a copy of the pendingFutures and use set::removeAll(<copy of pendingFutures>) to allow other send to take place while flushing.

Reply:
I think removing the specific future would involve having a mechanism to lookup that future base on the message id or someother id. I think the initial eventHubSystemProducer had that logic. I believe Daniel removed it and made it simpler. I like the current approach we have where we clear the pendingFutures on flush. 

flush is synchronized with send, So i don't think there should be race between them.

Reply:
You are using a set, you can simple use pendingFutures::remove as part of the handle (in method send). 

 sendResult.handle((aVoid, throwable) -> {
           pendingFutures.remove(sendResult);
       ....
}


===================
Another comment:
If a SystemProducer is not using the class NoFlushAsyncSystemProducer  (e.g. inherit directly from AsyncSystemProducer; I do not see how entries are removed from  pendingFutures.